File: integrations/alga-agent-sdk-js/src/commands.ts
// InvestigationCommand is the wire payload of a `kind: "inv_tool"` agent
// message. It maps 1:1 to the backend InvTool struct so the SDK can drive
// every command the backend supports. Fields are optional to keep payloads
// small. The incident identifier is `incident_number` (number), matching the
// backend contract — never `incident_id`.
export interface InvestigationCommand {
  op: string;
  // chat_id is the agent chat thread identifier. The backend grammar is
  // "alert_<number>", "incident_coord_<number>", "incident_inv_<number>".
  chat_id?: string;

  // --- Alert-scoped ---
  fingerprint?: string;

  // --- Generic mutation context ---
  reason?: string;
  note?: string;
  priority?: string;
  severity?: string;
  title?: string;

  // --- set_outcome / set_incident_resolution_docs / resolve_incident ---
  root_cause?: string;
  resolution?: string;
  summary?: string;
  impact_assessment?: string;
  actions_taken?: string;
  eta?: string;

  // --- triage_feedback ---
  triage_result_id?: string;
  agreed?: boolean;
  correct_decision?: string;
  correct_severity?: string;

  // --- assign_investigation / assign_incident_role ---
  target_agent_id?: string;
  role_type?: string;
  user_id?: string;
  agent_token_id?: string;
  scope_description?: string;

  // --- Incident-scoped (require incident_number) ---
  incident_number?: number;

  // --- coordination handoff / status ---
  message?: string;
  audience?: string;
  urgency?: string;
  status_level?: string;
  source_coordination_message_id?: string;
  internal?: boolean;
}

// --- Alert investigation tools ---

export function resolveAlert(fingerprint: string): InvestigationCommand {
  return { op: "resolve_alert", fingerprint };
}

export function reopenAlert(fingerprint: string): InvestigationCommand {
  return { op: "reopen_alert", fingerprint };
}

export function setOutcome(rootCause?: string, resolution?: string): InvestigationCommand {
  const cmd: InvestigationCommand = { op: "set_outcome" };
  if (rootCause !== undefined) cmd.root_cause = rootCause;
  if (resolution !== undefined) cmd.resolution = resolution;
  return cmd;
}

export function cancelInvestigation(reason?: string): InvestigationCommand {
  const cmd: InvestigationCommand = { op: "cancel_investigation" };
  if (reason !== undefined) cmd.reason = reason;
  return cmd;
}

export function pauseInvestigation(reason?: string): InvestigationCommand {
  const cmd: InvestigationCommand = { op: "pause_investigation" };
  if (reason !== undefined) cmd.reason = reason;
  return cmd;
}

export function triageFeedback(
  triageResultId: string,
  agreed = true,
  correctDecision?: string,
  correctSeverity?: string,
  note?: string,
): InvestigationCommand {
  const cmd: InvestigationCommand = {
    op: "triage_feedback",
    triage_result_id: triageResultId,
    agreed,
  };
  if (correctDecision !== undefined) cmd.correct_decision = correctDecision;
  if (correctSeverity !== undefined) cmd.correct_severity = correctSeverity;
  if (note !== undefined) cmd.note = note;
  return cmd;
}

export function assignInvestigation(targetAgentId: string): InvestigationCommand {
  return { op: "assign_investigation", target_agent_id: targetAgentId };
}

export function promoteToIncident(
  title?: string,
  severity?: string,
  priority?: string,
): InvestigationCommand {
  const cmd: InvestigationCommand = { op: "promote_to_incident" };
  if (title !== undefined) cmd.title = title;
  if (severity !== undefined) cmd.severity = severity;
  if (priority !== undefined) cmd.priority = priority;
  return cmd;
}

// --- Incident tools ---

export function setIncidentPriority(
  incidentNumber: number,
  priority: string,
): InvestigationCommand {
  return { op: "set_incident_priority", incident_number: incidentNumber, priority };
}

export function setIncidentSeverity(
  incidentNumber: number,
  severity: string,
): InvestigationCommand {
  return { op: "set_incident_severity", incident_number: incidentNumber, severity };
}

export function triggerEscalation(incidentNumber: number): InvestigationCommand {
  return { op: "trigger_escalation", incident_number: incidentNumber };
}

export function mitigateIncident(incidentNumber: number, reason?: string): InvestigationCommand {
  const cmd: InvestigationCommand = { op: "mitigate_incident", incident_number: incidentNumber };
  if (reason !== undefined) cmd.reason = reason;
  return cmd;
}

// resolveIncident is for commander-capable agents only. Investigator agents
// should record outcomes and ask the commander to verify resolution.
export function resolveIncident(incidentNumber: number, reason?: string): InvestigationCommand {
  const cmd: InvestigationCommand = { op: "resolve_incident", incident_number: incidentNumber };
  if (reason !== undefined) cmd.reason = reason;
  return cmd;
}

export function beginTriage(incidentNumber: number): InvestigationCommand {
  return { op: "begin_triage", incident_number: incidentNumber };
}

export function promoteIncident(incidentNumber: number): InvestigationCommand {
  return { op: "promote_incident", incident_number: incidentNumber };
}

// Assigns an incident role (e.g. "commander", "communicator") to a human user.
// The backend requires exactly one assignee, so this sets user_id only.
export function assignIncidentRoleToUser(
  incidentNumber: number,
  roleType: string,
  userId: string,
  scopeDescription?: string,
): InvestigationCommand {
  const cmd: InvestigationCommand = {
    op: "assign_incident_role",
    incident_number: incidentNumber,
    role_type: roleType,
    user_id: userId,
  };
  if (scopeDescription !== undefined) cmd.scope_description = scopeDescription;
  return cmd;
}

// Assigns an incident role to an agent token. The backend requires exactly one
// assignee, so this sets agent_token_id only.
export function assignIncidentRoleToAgent(
  incidentNumber: number,
  roleType: string,
  agentTokenId: string,
  scopeDescription?: string,
): InvestigationCommand {
  const cmd: InvestigationCommand = {
    op: "assign_incident_role",
    incident_number: incidentNumber,
    role_type: roleType,
    agent_token_id: agentTokenId,
  };
  if (scopeDescription !== undefined) cmd.scope_description = scopeDescription;
  return cmd;
}

// --- Coordination / status tools ---

// PostHandoff posts a coordination handoff message. Deprecated in favor of
// DispatchTask + CompleteTask — every handoff activates the commander and
// communicator and triggers ping-pong loops. Use CompleteTask to hand work
// back to the commander.
export function postHandoff(
  incidentNumber: number,
  message: string,
  audience: string,
  urgency: string,
): InvestigationCommand {
  return { op: "post_handoff", incident_number: incidentNumber, message, audience, urgency };
}

// PublishStatusUpdate publishes a status update. message and statusLevel are
// required. Status updates go to the Status Updates card and do NOT activate
// other agents — they are the safe channel for in-progress communication.
export function publishStatusUpdate(
  incidentNumber: number,
  message: string,
  statusLevel: string,
): InvestigationCommand {
  return {
    op: "publish_status_update",
    incident_number: incidentNumber,
    message,
    status_level: statusLevel,
  };
}

// SetIncidentResolutionDocs sets incident resolution documents. At least one
// of the optional fields must be provided.
export function setIncidentResolutionDocs(
  incidentNumber: number,
  opts: {
    summary?: string;
    impactAssessment?: string;
    actionsTaken?: string;
    rootCause?: string;
    resolution?: string;
  },
): InvestigationCommand {
  const cmd: InvestigationCommand = {
    op: "set_incident_resolution_docs",
    incident_number: incidentNumber,
  };
  if (opts.summary !== undefined) cmd.summary = opts.summary;
  if (opts.impactAssessment !== undefined) cmd.impact_assessment = opts.impactAssessment;
  if (opts.actionsTaken !== undefined) cmd.actions_taken = opts.actionsTaken;
  if (opts.rootCause !== undefined) cmd.root_cause = opts.rootCause;
  if (opts.resolution !== undefined) cmd.resolution = opts.resolution;
  return cmd;
}

File: integrations/alga-agent-sdk-go/commands.go
package alga

import "time"

// InvestigationCommand is the wire payload of a `kind: "inv_tool"` agent
// message. It maps 1:1 to the backend InvTool struct
// (apps/backend/api/agent/agent_tools.go) so the SDK can drive every command
// the backend supports. Fields are omitempty to keep payloads small.
type InvestigationCommand struct {
	Op string `json:"op"`
	// ChatID is the agent chat thread identifier. The backend grammar is
	// "alert_<number>" for alert investigations, "incident_coord_<number>"
	// for incident coordination threads, and "incident_inv_<number>" for
	// incident-scoped investigations. Most ops require this to be set so the
	// backend can authorize the caller against the underlying investigation.
	ChatID string `json:"chat_id,omitempty"`

	// --- Alert-scoped ---
	Fingerprint string `json:"fingerprint,omitempty"`

	// --- Generic mutation context ---
	Reason   string `json:"reason,omitempty"`
	Note     string `json:"note,omitempty"`
	Priority string `json:"priority,omitempty"`
	Severity string `json:"severity,omitempty"`
	Title    string `json:"title,omitempty"`

	// --- set_outcome ---
	RootCause        *string `json:"root_cause,omitempty"`
	Resolution       *string `json:"resolution,omitempty"`
	Summary          string  `json:"summary,omitempty"`
	ImpactAssessment string  `json:"impact_assessment,omitempty"`
	ActionsTaken     string  `json:"actions_taken,omitempty"`
	ETADetail        string  `json:"eta,omitempty"`

	// --- triage_feedback ---
	TriageResultID  string `json:"triage_result_id,omitempty"`
	Agreed          bool   `json:"agreed,omitempty"`
	CorrectDecision string `json:"correct_decision,omitempty"`
	CorrectSeverity string `json:"correct_severity,omitempty"`

	// --- assign_investigation / assign_incident_role ---
	TargetAgentID    string  `json:"target_agent_id,omitempty"`
	RoleType         string  `json:"role_type,omitempty"`
	UserID           string  `json:"user_id,omitempty"`
	AgentTokenID     string  `json:"agent_token_id,omitempty"`
	ScopeDescription *string `json:"scope_description,omitempty"`

	// --- Incident-scoped (require IncidentNumber) ---
	IncidentNumber int64 `json:"incident_number,omitempty"`

	// --- coordination handoff / status ---
	Message                     string `json:"message,omitempty"`
	Audience                    string `json:"audience,omitempty"`
	Urgency                     string `json:"urgency,omitempty"`
	StatusLevel                 string `json:"status_level,omitempty"`
	SourceCoordinationMessageID string `json:"source_coordination_message_id,omitempty"`
	Internal                    bool   `json:"internal,omitempty"`
}

// StrPtr is a convenience helper for the *string command fields.
func StrPtr(s string) *string { return &s }

// --- Alert investigation tools ---

func ResolveAlert(fingerprint string) InvestigationCommand {
	return InvestigationCommand{Op: "resolve_alert", Fingerprint: fingerprint}
}

func ReopenAlert(fingerprint string) InvestigationCommand {
	return InvestigationCommand{Op: "reopen_alert", Fingerprint: fingerprint}
}

func SetOutcome(rootCause, resolution *string) InvestigationCommand {
	return InvestigationCommand{Op: "set_outcome", RootCause: rootCause, Resolution: resolution}
}

func CancelInvestigation(reason string) InvestigationCommand {
	return InvestigationCommand{Op: "cancel_investigation", Reason: reason}
}

func PauseInvestigation(reason string) InvestigationCommand {
	return InvestigationCommand{Op: "pause_investigation", Reason: reason}
}

func TriageFeedback(triageResultID string, agreed bool, correctDecision, correctSeverity, note string) InvestigationCommand {
	return InvestigationCommand{
		Op:              "triage_feedback",
		TriageResultID:  triageResultID,
		Agreed:          agreed,
		CorrectDecision: correctDecision,
		CorrectSeverity: correctSeverity,
		Note:            note,
	}
}

func AssignInvestigation(targetAgentID string) InvestigationCommand {
	return InvestigationCommand{Op: "assign_investigation", TargetAgentID: targetAgentID}
}

func PromoteToIncident(title, severity, priority string) InvestigationCommand {
	return InvestigationCommand{Op: "promote_to_incident", Title: title, Severity: severity, Priority: priority}
}

// --- Incident tools ---

func SetIncidentPriority(incidentNumber int64, priority string) InvestigationCommand {
	return InvestigationCommand{Op: "set_incident_priority", IncidentNumber: incidentNumber, Priority: priority}
}

func SetIncidentSeverity(incidentNumber int64, severity string) InvestigationCommand {
	return InvestigationCommand{Op: "set_incident_severity", IncidentNumber: incidentNumber, Severity: severity}
}

func TriggerEscalation(incidentNumber int64) InvestigationCommand {
	return InvestigationCommand{Op: "trigger_escalation", IncidentNumber: incidentNumber}
}

func MitigateIncident(incidentNumber int64, reason string) InvestigationCommand {
	return InvestigationCommand{Op: "mitigate_incident", IncidentNumber: incidentNumber, Reason: reason}
}

// ResolveIncident is for commander-capable agents only. Investigator agents
// should record outcomes and ask the commander to verify resolution through
// coordination.
func ResolveIncident(incidentNumber int64, reason string) InvestigationCommand {
	return InvestigationCommand{Op: "resolve_incident", IncidentNumber: incidentNumber, Reason: reason}
}

func BeginTriage(incidentNumber int64) InvestigationCommand {
	return InvestigationCommand{Op: "begin_triage", IncidentNumber: incidentNumber}
}

func PromoteIncident(incidentNumber int64) InvestigationCommand {
	return InvestigationCommand{Op: "promote_incident", IncidentNumber: incidentNumber}
}

// AssignIncidentRoleToUser assigns an incident role (e.g. "commander",
// "communicator") to a human user. The backend requires exactly one assignee,
// so this sets user_id only. Pass scopeDescription "" to omit it.
func AssignIncidentRoleToUser(incidentNumber int64, roleType, userID, scopeDescription string) InvestigationCommand {
	cmd := InvestigationCommand{
		Op:             "assign_incident_role",
		IncidentNumber: incidentNumber,
		RoleType:       roleType,
		UserID:         userID,
	}
	if scopeDescription != "" {
		cmd.ScopeDescription = &scopeDescription
	}
	return cmd
}

// AssignIncidentRoleToAgent assigns an incident role to an agent token. The
// backend requires exactly one assignee, so this sets agent_token_id only.
// Pass scopeDescription "" to omit it.
func AssignIncidentRoleToAgent(incidentNumber int64, roleType, agentTokenID, scopeDescription string) InvestigationCommand {
	cmd := InvestigationCommand{
		Op:             "assign_incident_role",
		IncidentNumber: incidentNumber,
		RoleType:       roleType,
		AgentTokenID:   agentTokenID,
	}
	if scopeDescription != "" {
		cmd.ScopeDescription = &scopeDescription
	}
	return cmd
}

// --- Coordination / status tools ---

// PostHandoff posts a coordination handoff message. message, audience, and
// urgency are required by the backend. The backend accepts only the
// structured responder→commander handoff and urgent commander-attention
// calls; see the incident role prompts for the expected message format.
func PostHandoff(incidentNumber int64, message, audience, urgency string) InvestigationCommand {
	return InvestigationCommand{
		Op:             "post_handoff",
		IncidentNumber: incidentNumber,
		Message:        message,
		Audience:       audience,
		Urgency:        urgency,
	}
}

// PublishStatusUpdate publishes a status update. message and statusLevel are
// required. Status updates go to the Status Updates card and do NOT activate
// other agents — they are the safe channel for in-progress communication.
func PublishStatusUpdate(incidentNumber int64, message, statusLevel string) InvestigationCommand {
	return InvestigationCommand{
		Op:             "publish_status_update",
		IncidentNumber: incidentNumber,
		Message:        message,
		StatusLevel:    statusLevel,
	}
}

// SetIncidentResolutionDocs sets incident resolution documents. At least one
// of the optional fields must be provided.
func SetIncidentResolutionDocs(incidentNumber int64, summary, impactAssessment, actionsTaken, rootCause, resolution string) InvestigationCommand {
	cmd := InvestigationCommand{
		Op:               "set_incident_resolution_docs",
		IncidentNumber:   incidentNumber,
		Summary:          summary,
		ImpactAssessment: impactAssessment,
		ActionsTaken:     actionsTaken,
	}
	if rootCause != "" {
		cmd.RootCause = &rootCause
	}
	if resolution != "" {
		cmd.Resolution = &resolution
	}
	return cmd
}

// --- Time helper ---

// Now returns the current time. Wrapped as a package var so tests can stub
// deterministic timestamps (e.g. for idempotency-key generation).
var Now = func() time.Time { return time.Now() }
